File: ruby/difference-of-squares/difference_of_squares.rb
class Squares
  SQUARE = 2
  private_constant :SQUARE

  def initialize(number)
    @sum = (1..number).method(:sum)
  end

  def square_of_sum
    sum.call**SQUARE
  end

  def sum_of_squares
    sum.call { |number| number**SQUARE }
  end

  def difference
    square_of_sum - sum_of_squares
  end

  private

  attr_reader :sum
end

module BookKeeping
  VERSION = 4
end

File: javascript/difference-of-squares/difference-of-squares.js
export class Squares {
  static get SQUARE() {
    return 2
  }

  constructor(number) {
    this._number = number
  }

  get number() {
    return this._number
  }

  get sumOfSquares() {
    return this.numberRange().reduce(this.addSquare(), 0)
  }

  get squareOfSum() {
    return this.numberRange().reduce(this.sum, 0)**Squares.SQUARE
  }

  get difference() {
    return this.squareOfSum - this.sumOfSquares
  }

  numberRange() {
    return [...Array(this.number).keys()].map(this.increment())
  }

  addSquare() {
    return (acc, number) => {
      return this.sum(acc, number**Squares.SQUARE)
    }
  }

  increment() {
    return (number) => {
      return this.sum(1, number)
    }
  }

  sum(acc, number) {
    return acc + number
  }
}
